src/network/networkd-network.c

```diff
@@ -979,12 +980,18 @@ int config_parse_hostname(
 
         if (!hostname_is_valid(hn, false)) {
                 log_syntax(unit, LOG_ERR, filename, line, 0, "Hostname is not valid, ignoring assignment: %s", rvalue);
-                free(hn);
+                return 0;
+        }
+
+        if (dns_name_is_valid(hn) <= 0) {
+                log_syntax(unit, LOG_ERR, filename, line, 0, "Hostname is not a valid DNS domain name, ignoring assignment: %s", rvalue);
                 return 0;
         }
 
         free(*hostname);
         *hostname = hostname_cleanup(hn);
```

Comment:
Hmm, calling `hostname_cleanup` after validation seems wrong. Shouldn't it be the other way around?